sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -164,146 +152,121 @@ public void forceFlush() {
+      flushRequested = null;
+    }
+
+    private void updateNextExportTime() {
+      nextExportTime = System.nanoTime() + scheduleDelayNanos;
     }
 
     private void shutdown() {
       forceFlush();
-      timer.cancel();
       spanExporter.shutdown();
+      continueWork = false;
     }
 
     private void forceFlush() {
-      ArrayList<ReadableSpan> spansCopy;
-      synchronized (monitor) {
-        spansCopy = new ArrayList<>(spansList);
-        spansList.clear();
+      CompletableResultCode flushResult = new CompletableResultCode();
+      this.flushRequested = flushResult;
+      try {
+        flushResult.join();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
       }
-      // Execute the batch export outside the synchronized to not block all producers.
-      exportBatches(spansCopy);
     }
 
-    private void exportBatches(ArrayList<ReadableSpan> spanList) {
-      // TODO: Record a counter for pushed spans.
-      for (int i = 0; i < spanList.size(); ) {
-        int lastIndexToTake = Math.min(i + maxExportBatchSize, spanList.size());
-        onBatchExport(createSpanDataForExport(spanList, i, lastIndexToTake));
-        i = lastIndexToTake;
+    private void exportCurrentBatch() {
+      if (batch.isEmpty()) {
+        return;
       }
-    }
 
-    private static List<SpanData> createSpanDataForExport(
-        List<ReadableSpan> spanList, int startIndex, int endIndex) {
-      List<SpanData> spanDataBuffer = new ArrayList<>(endIndex - startIndex);
-      for (int i = startIndex; i < endIndex; i++) {
-        spanDataBuffer.add(spanList.get(i).toSpanData());
-        // Remove the reference to the ReadableSpan to allow GC to free the memory.
-        spanList.set(i, null);
-      }
-      return Collections.unmodifiableList(spanDataBuffer);
-    }
-
-    // Exports the list of SpanData to the SpanExporter.
-    @SuppressWarnings("BooleanParameter")
-    private void onBatchExport(final List<SpanData> spans) {
-      if (exportAvailable.compareAndSet(true, false)) {
-        try {
-          final CompletableResultCode result = spanExporter.export(spans);
-          result.whenComplete(
-              new Runnable() {
-                @Override
-                public void run() {
-                  if (!result.isSuccess()) {
-                    logger.log(Level.FINE, "Exporter failed");
-                  }
-                  exportAvailable.set(true);
-                }
-              });
-          timer.schedule(
-              new TimerTask() {
-                @Override
-                public void run() {
-                  result.fail();
-                }
-              },
-              exporterTimeoutMillis);
-        } catch (Exception e) {
-          logger.log(Level.WARNING, "Exporter threw an Exception", e);
+      try {
+        final CompletableResultCode result = spanExporter.export(batch);
+        result.join(exporterTimeoutMillis, TimeUnit.MILLISECONDS);
```

Comment:
I am not sure about that. We have to prevent concurrent calls to export and we have to have timeouts. IMO timed joins is the easiest way to achieve this. At least until benchmarking shows that this is too expensive.

Reply:
Join shouldn’t have a timeout. There’s nothing you can really do if you can’t join. This is why CompleteableFuture and even back to pthreads has no join.

If an exporter doesn’t return then there is nothing that can be done here other than fail it’s result. Exporters can then act on a result failure and halt their activity.